### ISSUE TITLE
casc reload groovy command breaks jobdsl plugin if executed by pipeline step load()

### ISSUE THREAD

**Issue by aliusmiles** [jenkinsci/configuration-as-code-plugin#1227]
### Your checklist for this issue

- [x] Jenkins version: 2.190.3

- [x] Plugin version: 1.34

- [x] OS: docker image based on jenkins/jenkins:lts

### Description

If jobs section is present in jenkins.yaml, configuration reload fails, both via UI and groovy script. Jobs config works without any issues on init, but fails on reload. CASC_JENKINS_CONFIG is set to url.

Config example:
```
jobs:
  - script: >
      folder('xxx')
```

Error: ```groovy.lang.MissingMethodException: No signature of method: javaposse.jobdsl.plugin.JenkinsDslScriptLoader.customizeCompilerConfiguration() is applicable for argument types: (org.codehaus.groovy.control.CompilerConfiguration) values: [org.codehaus.groovy.control.CompilerConfiguration@67233c11]```

Full trace: https://pastebin.com/0A2Mtz8z

**Comment by casz** [comment]
what version of jobdsl?

**Comment by aliusmiles** [comment]
1.76

**Comment by aliusmiles** [comment]
Update: 

issue is reproducing only if reload using declarative pipeline step load:

reload.groovy file:
```
import io.jenkins.plugins.casc.ConfigurationAsCode;
ConfigurationAsCode.get().configure()
```
pipeline step:
```
load('reload.groovy')
```

UI, token, script console and freestyle job with 'execute system groovy script' are working just fine.
And after running declarative step above, all other reload methods start failing as well.

Are there any examples on how to trigger casc reload from declarative pipeline? I wasn't able to find any..

**Comment by aliusmiles** [comment]
For reference, I was able to trigger casc reload from declarative pipeline by creating via jobdsl freestyle job with systemGroovyCommand step and then invoking that job by *build* step.

Would be nice to trigger reload directly from declarative pipeline though.

**Comment by aliusmiles** [comment]
still the same error with freestyle reload job: it runs just fine immediately after jenkins start, but after running jobdsl step in another job reload job starts failing and jobdsl step as well. It seems to me that groovy code for reload is somehow breaking jobdsl plugin.

**Comment by casz** [comment]
See https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/features/configurationReload.md

**Comment by aliusmiles** [comment]
To be honest, that page is quite lacking some important details. Like example of Jenkins CLI command or mention that only system groovy script will work.. Should I maybe just create PR for that? :)

I was able to find the root of this issue: forgot to remove `load('reload.groovy')` step from test job created by jobdsl 🤦🏻‍♂️ so seed job and reload job were working correctly but then test job was built by branch indexing and breaking jobdsl plugin..

Conclusion: using `load` pipeline step to run casc reload groovy command effectively breaks jobdsl plugin. Guess it's a lack of permissions, but quite weird that it's affecting jobdsl plugin in such way.


P.S. Maybe it will be useful for someone, my configs for casc reload:

jobdsl for reload job:
```
job('casc-reload') {
    steps {
        systemGroovyCommand("""
            import io.jenkins.plugins.casc.ConfigurationAsCode;
            ConfigurationAsCode.get().configure()
        """.stripIndent()) {
            sandbox(true)
        }
    }
}
```
declarative pipeline step to trigger reload:
```
build('casc-reload')
```
required permissions in casc:

```
  scriptApproval:
    approvedSignatures:
      - 'staticMethod io.jenkins.plugins.casc.ConfigurationAsCode get'
      - 'method io.jenkins.plugins.casc.ConfigurationAsCode configure'
```

**Comment by casz** [comment]
Feel free to contribute a PR to improve the docs

**Comment by hak33m16** [comment]
For anybody else having issues getting @aliusmiles recommended script to work, make sure that you have the Groovy plugin installed. Interestingly, the JobDSL plugin doesn't seem to fail, it'll just create an empty freestyle job, but if you check the logs, you should see something like:
```
Warning: (script, line 2055) version 2.0 or later of plugin 'groovy' needs to be installed
```

**Comment by MarkEWaite** [comment]
> Hi so is the verdict that we cannot reload using a declarative pipeline or something? Any reasons behind why? Can it be addressed? I've just ran into the same issue as I'm trying to migrate stuff to GitHub branch source plugin where possible. Any thoughts on how to do this as part of a declarative pipeline?
> 
> As it stands I've created a new file `vars/reloadjcasc.groovy` in my shared library with the groovy in that page, and simply run `reloadjcasc` in my `Jenkinsfile` but then this error happens.

Reloading JCasC is a system level operation.  Allowing that from declarative pipeline risks all sorts of security mischief.  I would not expect it to be allowed from declarative pipeline any more than I would expect that declarative pipeline would ever be allowed to change the system security realm.

**Comment by hak33m16** [comment]
> Reloading JCasC is a system level operation. Allowing that from declarative pipeline risks all sorts of security mischief. I would not expect it to be allowed from declarative pipeline any more than I would expect that declarative pipeline would ever be allowed to change the system security realm.

Do you have any recommendations on how to automatically reload it securely then?

I _think_ this is all really just working around the fact that JCaC offers no native way to _automatically_ sync configuration from an external (preferably private) repository. The lack of that makes it largely useless IMO, because users have to, each time they make changes, go to the JCasC page and point it to the correct location to pull configuration from. And from looking, I couldn't personally find a way to POST the JCasC in any officially documented way. Paired with the difficulty involved in figuring out what to put in there in the first place, it hardly feels worth the effort of using

Maybe this rant would be best suited to exist in a separate feature request for JCasC, or someone could point out to my why/how the suggestion is insecure

**Comment by jetersen** [comment]
There are certainly ways and they are documented.

https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/features/configurationReload.md

**Comment by MarkEWaite** [comment]
> There are certainly ways and they are documented.
> 
> https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/features/configurationReload.md

I confirmed that at least one of the documented techniques works on my Jenkins instance.  I ran the command:

```
java -jar jenkins-cli.jar -auth @mwaite-auth -s http://mark-pc2.markwaite.net:8080/ reload-jcasc-configuration
```

that reads my credentials (username and a Jenkins generated authentication token) from the file `mwaite-auth` and reloads the configuration.  In my case, my static agents are defined with JCasC.  I added a label to one of the agents, reloaded the configuration from that command line call, and confirmed that the label I had just added was no longer there.

**Comment by fugkco** [comment]
> Reloading JCasC is a system level operation. Allowing that from declarative pipeline risks all sorts of security mischief. I would not expect it to be allowed from declarative pipeline any more than I would expect that declarative pipeline would ever be allowed to change the system security realm.

I guess my thinking is that running any of those documented methods (including OP's workaround) from a pipeline defeats the security model as far as I understand. It _feels_ like using the API or the CLI is even worse as it requires adding credentials to Jenkins that need system administer access (please correct me if I'm wrong). Surely allowing a pipeline to _only_ reload the configuration (without additional changes) is non-destructive. If the pipeline doesn't have access to make changes to file system, then it is futile to invoke a reload as it will just load up what was already there.

In our case, we are already trusting our Jenkins repo to update the jobs, and it also contains our full infrastructure for Jenkins. Perhaps what is needed is a pipeline step to do a reload that doesn't break the security model, e.g. it requires running the pipeline as a user or something (I don't know if such functionality exists for declarative pipelines). Or better yet, maybe a new permission can be introduced for reloading configuration. The latter would be nice as it allows creation of a user that _only_ has access to that, thus limiting the blast radius.

Regardless of all of that, there is a bug that in my opinion needs fixing. Running `ConfigurationAsCode.get().configure()` from a declarative pipeline seems to put Jenkins in a bad state where this plugin cannot be reloaded until after a restart (with the stack trace in OP's link) regardless of who does it or what mechanisms is used. As an example, in my local environment I'm admin and after running the code in a declarative pipeline once, I'm unable to reload it at all even if I go into the administer UI, or use the API, both with valid credentials and permissions. I'm raising a PR after this comment to document this in the config reload docs, but it would be nice to have a proper fix in place. If you like, I can raise a separate issue to address this.

**Comment by timja** [comment]
There's a rest API for doing this with just a pre-shared token that doesn't require administrator privileges.

> via http POST to JENKINS_URL/reload-configuration-as-code

**Comment by fugkco** [comment]
Yeah I saw that, but personally if it's a problem to kick off a reload from a pipeline, allowing anyone (almost) without auth to kick off a reload is an even bigger bigger issue. I've already settled on that as there's no other way it seems, but like I said, ideally there would be a way to kick off a reload directly from a build. Perhaps this would be a job that needs to be allowed or something but still.

Either way, I think fixing the issue where kicking off a reload from a declarative pipeline causes Jenkins/JCASC to be in a bad state would be great.
